---
title: "fourcore_subtype_annotation"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../"))
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

```{r library, include=FALSE}

setwd(paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../"))

library(scran)
library(scater)
library(ggplot2)
library(bsub)
library(ComplexHeatmap)
library(tidyverse)
library(celldex)
library(SingleR)

# needed to run quasr
bsub_opt$call_Rscript = function(version) {
  GetoptLong::qq("module load libpng/1.6.37, module load R/@{version}; Rscript")
}

bsub_opt$user <- "panten"

```

This script performs all analysis regarding the 4Core genotype mice -- verification of genotyping, deg analysis, celltype abundance

```{r load and prepare the data}

sce.aggregate <- readRDS("./processed_data/fourcore_sce_aggregate_across_samples.rds")

```


```{r deg_analysis for autosome and Y chromosome}

# degs for x, y and autosome chromosomes 

library(DESeq2)

make_ma_plot_all_genes <- function(deseq_dataset){
  deseq_dataset <- estimateSizeFactors(deseq_dataset)
  deseq_dataset <- DESeq(deseq_dataset)
  test <- results(deseq_dataset)
  data.frame(head(test[order(test$padj), ], n = 20))
  
  rowData(sce.aggregate) %>% data.frame() %>% rownames_to_column("Gene") %>% pull(chromosome, name = Gene) -> gene_to_chromosome
  
  test %>% data.frame() %>% 
    rownames_to_column("Gene") %>%
    add_column("Chromosome" = gene_to_chromosome[.$Gene]) %>%
    # dplyr::filter(Chromosome == "X") %>%
    dplyr::filter(baseMean > 50) %>%
    mutate(padj = p.adjust(pvalue)) %>%
    mutate(labelling = ifelse(padj > .1, "grey", case_when(
      Chromosome == "X" ~ "X", 
      Chromosome == "Y" ~ "Y", 
      .default = "Autosome"
    ))) %>%
    mutate(FC = ifelse(abs(log2FoldChange) > 2, sign(log2FoldChange) * 2, log2FoldChange)) -> test_df
  
  test_df %>%
    dplyr::filter(baseMean > 10) %>% 
    {
      ggplot(., aes(x = baseMean, y = FC, fill = labelling)) + geom_point(pch = 21, size = 2) + scale_x_log10() + 
        # ggrepel::geom_text_repel(data = . %>% dplyr::filter(padj < .1), aes(label = Gene)) + 
        ggrepel::geom_text_repel(data = . %>% dplyr::filter(padj < .1), aes(label = Gene), size = 8) + 
        geom_hline(yintercept = 1, linetype = 'dashed') + 
        scale_fill_manual(values = c("grey" =  "grey", "X" =  "red", Y = "green","Autosome" = "blue")) + 
        theme_bw(base_size = 30) + xlab("Expression level") + ylab("XYT / XXT (log2 fold change)") + 
        theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_rect(fill=NA, colour = "black", size=1)) +
        theme(legend.position = "None")
  }
}

make_ma_plot_all_genes(DESeqDataSet(sce.aggregate[,sce.aggregate$Gonadal == "female"], design = ~ Library + Genotype)) + ylab("XYO / XXO (log2 fold change)")
ggplot2::ggsave("./plots/scrnaseq/full_spleen_gonadal_female_autosomes.pdf", width = 8, height = 8)

make_ma_plot_all_genes(DESeqDataSet(sce.aggregate[,sce.aggregate$Gonadal == "male"], design = ~ Library + Genotype)) + ylab("XYT / XXT (log2 fold change)")
ggplot2::ggsave("./plots/scrnaseq/full_spleen_gonadal_male_autosomes.pdf", width = 8, height = 8)

```
