---
title: "Celltype_graph_paper"
output: html_document
date: '2023-11-01'
editor_options: 
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../"))
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```


```{r library, include=FALSE}

setwd(paste0(dirname(rstudioapi::getSourceEditorContext()$path), "/../"))

library(scran)
library(scater)
library(ggplot2)
library(Rtsne)
library(irlba)
library(RColorBrewer)
library(viridis)
library(pheatmap)
library(rtracklayer)
library(gridExtra)
library(writexl)
library(PCAtools)
library(BiocSingular)
library(scuttle)
library(batchelor)
library(org.Mm.eg.db)
library(limma)
library(celldex)
library(bluster)
library(edgeR)
library(bsub)
library(ComplexHeatmap)
library(tidyverse)
library(celldex)
library(SingleR)

# needed to run quasr
bsub_opt$call_Rscript = function(version) {
  GetoptLong::qq("module load libpng/1.6.37, module load R/@{version}; Rscript")
}

```


```{r load and prepare the data}
# load data
sce <- readRDS("./processed_data/fourcore_sce.final.rds")
meta <- readRDS("./processed_data/fourcore_sce.final_meta.rds")

# Check number of cells and samples
dim(sce) ##103220 cells 
length(table(sce$Sample)) ##16 samples

# add genotype information
genotype_metadata <- read_csv("./metadata_rna_four_core.csv") %>% 
  mutate(sample_identifier = paste0(seq_ids, "_", ILSE_NO))

genotype_annotation <- genotype_metadata %>% pull(GENOTYPE, name = sample_identifier)

sce$Genotype <- genotype_annotation[sce$Sample]

# clusters out: -- likely doublets, outliers
clusters_out <- c(14, 26, 27)
sce <- sce[,!sce$label %in% clusters_out]

# aggregate across samples
sce.aggregate <- aggregateAcrossCells(sce, DataFrame(Sample = sce$Sample, Library = sce$Library))

# add chromosome annotation to the sce.aggregate
library('AnnotationHub')
ah <- AnnotationHub()
query(ah, c("Mus musculus", "Ensembl", "v97"))
ensdb <- ah[["AH73905"]]

chromosome <- mapIds(ensdb,
    keys = rownames(sce.aggregate),
    keytype = "SYMBOL",
    column = "SEQNAME")
rowData(sce.aggregate)$chromosome <- chromosome

sce.aggregate <- sce.aggregate[!is.na(rowData(sce.aggregate)$chromosome), ]
```


```{r celltype_distribution}
### we correct the one sample that is swapped: 
swapped_sample1 <- "87884148_34406" # annotated as XYF, actually XXF
swapped_sample2 <- "87882871_34406" # annotated as XXF, actually XYF

sce[,sce$Sample == swapped_sample1]$Genotype <- "F1_FCG_XXF"
sce[,sce$Sample == swapped_sample2]$Genotype <- "F1_FCG_XYF"

samples.test <- c("90130514_36306"="XXO-F1-01", "93202067_36145"="XXO-F1-02", "87884148_34406"="XXO-F1-03", "91951650_35978"="XXO-F1-04",#XXF
             "90131144_36306"="XXT-F1-01","93202705_36145"="XXT-F1-02", "87883510_34406"="XXT-F1-03", "91952280_35978"="XXT-F1-04", #XXM
        "90131774_36306"="XYO-F1-01", "93203343_36145"="XYO-F1-02", "87882871_34406"="XYO-F1-03", "91952910_35978"="XYO-F1-04", # XYF
        "90132404_36306"="XYT-F1-01", "93203981_36145"="XYT-F1-02", "87884793_34406"="XYT-F1-03", "91953540_35978"="XYT-F1-04") #XYM

samples <- c("90130514_36306", "93202067_36145","87884148_34406","91951650_35978", #XXF
             "90131144_36306","93202705_36145", "87883510_34406", "91952280_35978", #XXM
        "90131774_36306", "93203343_36145","87882871_34406", "91952910_35978", # XYF
        "90132404_36306", "93203981_36145", "87884793_34406", "91953540_35978") #XYM

#Plot cells proportion per each sample
colData(sce) %>% 
  data.frame() %>% 
  ggplot(aes(x = Sample, fill = AnnotatedCellType)) + geom_bar(pos = "fill", col = "black", width=0.5) + 
  scale_x_discrete(limits=samples, labels=samples.test, expand = c(0.03,0)) +
  scale_y_continuous(expand = c(0,0)) +
  labs(x="",y="cell type %") + 
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line.y.left  = element_line(color = 'black', size=1), panel.border = element_blank()) +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1, size = 8), axis.text.y = element_text(size = 8))+
  theme(text=element_text(family="sans"))
ggsave("./plots/scrnaseq/cells_proportion.pdf", width = 6, height = 3)
```



```{r Plot TSNE}

data.frame(
  Tsne1 = reducedDims(sce)[["TSNE"]][,1], 
  Tsne2 = reducedDims(sce)[["TSNE"]][,2], 
  CellType = sce$AnnotatedCellType,
  Sample = sce$Sample, 
  Genotype = sce$Genotype
) %>% 
  ggplot(aes(x = Tsne1, y = Tsne2, col = CellType)) +
    geom_point(size = 0.1) + 
    theme_bw() + 
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_rect(fill=NA, colour = "black", size=1), axis.ticks.y=element_blank(), axis.ticks.x=element_blank()) +
    labs(x="",y="")+
    theme(axis.text.x =element_blank() , axis.text.y = element_blank(), plot.title = element_text(face = "bold"), legend.position="none")+
  theme(text=element_text(family="sans"))
ggsave("./plots/scrnaseq/umap_spleen.png", width = 6, height = 4, dpi=700, bg="transparent")

```


```{r Plot markers genes}
#Plot markers genes for rest cells separately - for supplemental
markers <- c("Cd79b", "Ebf1", "Mzb1", "Vpreb3", "Ighd", # B cells
             "Skap1", "Lef1", "Cd3g",   #CD4naive
             "Cd8b1","Dapl1","Cd4", #CD8naive
             "Izumo1r",
             "Nkg7","Gzma", "Il2rb", #NKcells
             "Ccl4", "Ccl3", # Basophils
             "S100a8", "Il1b",    # Neutrophils
             "Ifitm3" #Monocyte
             )

cells <- c(
"B-cell (Fo)",       
"B-cell (Mz)",    
"B-cell (pre)",           
"B-cell (T)",             
"T-cell (CD4 control)",
"T-cell (CD8 naive)",
"T-cell (CD4 naive)",   
"T-cell (regulatory)",
"T-cell (gamma-delta)",
"NK cells",
"Basophil",           
"Neutrophils",
"Monocytes"
)

cell.label <- c(
"B-cell (Fo)"="B-Fo",      
"B-cell (Mz)"="B-Mz",
"B-cell (pre)"="B-pre",           
"B-cell (T)"="B-Tran",  
"T-cell (CD4 control)"="CD4ctrl",
"T-cell (CD8 naive)"="CD8naiv",
"T-cell (CD4 naive)"="CD4naiv", 
"T-cell (regulatory)"="T-reg",
"T-cell (gamma-delta)"="T-gd",
"NK cells"="NK",
"Basophil"="Baso",         
"Neutrophils"="Neutro",
"Monocytes"="Mono"
)

  
bw_palette <- c( "white", "black")

plotDots(sce, features = markers, group="AnnotatedCellType", color = c(bw_palette), zlim = c(0,4), exprs_values = "logcounts") + 
  scale_y_discrete(limits=markers) +  
  scale_x_discrete(limits=cells, labels=cell.label) +
  theme_bw() + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_rect(fill=NA, colour = "black", size=1)) +
  labs(x="",y="") +
  theme(axis.text.x = element_text(angle = 60, vjust = 1, hjust=1, size = 8), axis.text.y = element_text(size = 8), plot.title = element_text(face = "bold")) +
  theme(text=element_text(family="sans")) + 
  scale_fill_manual(values = bw_palette)
ggsave("./plots/scrnaseq/markers_genes.pdf", width = 5, height = 5)

# Check expression patterns per cluster & cell type
tcells.markers <- findMarkers(sce, sce$AnnotatedCellType)
data_results <- data.frame(tcells.markers[["T-cell (regulatory)"]])
data_results <- data_results[!grepl("Rp", rownames(data_results)), ]
head(data_results[order(-data_results$summary.logFC), 1:4], n = 30)
```



